packages/kernel-6.1/1006-strscpy-write-destination-buffer-only-once.patch

```diff
@@ -0,0 +1,128 @@
+Date: Sun, 1 Dec 2024 09:23:33 -0800
+Subject: [PATCH] strscpy: write destination buffer only once
+
+commit 9022ed0e7e65734d83a0648648589b9fbea8e8c9 upstream.
+
+The point behind strscpy() was to once and for all avoid all the
+problems with 'strncpy()' and later broken "fixed" versions like
+strlcpy() that just made things worse.
+
+So strscpy not only guarantees NUL-termination (unlike strncpy), it also
+doesn't do unnecessary padding at the destination.  But at the same time
+also avoids byte-at-a-time reads and writes by _allowing_ some extra NUL
+writes - within the size, of course - so that the whole copy can be done
+with word operations.
+
+It is also stable in the face of a mutable source string: it explicitly
+does not read the source buffer multiple times (so an implementation
+using "strnlen()+memcpy()" would be wrong), and does not read the source
+buffer past the size (like the mis-design that is strlcpy does).
+
+Finally, the return value is designed to be simple and unambiguous: if
+the string cannot be copied fully, it returns an actual negative error,
+making error handling clearer and simpler (and the caller already knows
+the size of the buffer).  Otherwise it returns the string length of the
+result.
+
+However, there was one final stability issue that can be important to
+callers: the stability of the destination buffer.
+
+In particular, the same way we shouldn't read the source buffer more
+than once, we should avoid doing multiple writes to the destination
+buffer: first writing a potentially non-terminated string, and then
+terminating it with NUL at the end does not result in a stable result
+buffer.
+
+Yes, it gives the right result in the end, but if the rule for the
+destination buffer was that it is _always_ NUL-terminated even when
+accessed concurrently with updates, the final byte of the buffer needs
+to always _stay_ as a NUL byte.
+
+[ Note that "final byte is NUL" here is literally about the final byte
+  in the destination array, not the terminating NUL at the end of the
+  string itself. There is no attempt to try to make concurrent reads and
+  writes give any kind of consistent string length or contents, but we
+  do want to guarantee that there is always at least that final
+  terminating NUL character at the end of the destination array if it
+  existed before ]
+
+This is relevant in the kernel for the tsk->comm[] array, for example.
+Even without locking (for either readers or writers), we want to know
+that while the buffer contents may be garbled, it is always a valid C
+string and always has a NUL character at 'comm[TASK_COMM_LEN-1]' (and
+never has any "out of thin air" data).
+
+So avoid any "copy possibly non-terminated string, and terminate later"
+behavior, and write the destination buffer only once.
+
+Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
+[vyaghras:
+  - backport to 6.1
+  - apply changes in original patch to strscpy, since sized_strscpy
+    doesn't exist in 6.1]
```

Comment:
Per the example commit I linked to, and my earlier offline suggestion, it might be more conventional to write this note like this:
```
[vyaghras: adjust context since e6584c3964f2f ("string: Allow 2-argument
strscpy()") is not present]
```